salt/_modules/metalk8s_blkid.py

```diff
@@ -0,0 +1,140 @@
+# coding: utf-8
+
+"""Wrapper exposing a Python-like interface around the libblkid functions.
+
+Why this module instead of shelling out to `blkid`?
+===================================================
+
+Because we want to be as sure as possible that when `blkid` returns no
+informationg is means "there is no info" and not "maybe there is some info but I
+couldn't get them" (because if we mix two cases, we may actually format
+something because we thought "ho there is nothing, davai" whereas the reality is
+"there was data but I haven't seen them because reasons…").
+
+Relying on `return code 2 == no data` seems kinda safe, but it's not 100% clear
+from the man page. Now we could check the source code (and I did) but that would
+only be valid for the current version we look at.
+
+So let's directly attack the library, making sure that any API call will result
+in an exception and thus if we end up with "no data" then that truly means
+"everything went well, but there was nothing there".
+"""
+
+import collections
+import ctypes
+import functools
+
+import metalk8s_blkid_c as blkid_c
+
+
+# No enum in Python 2 :-(
+class SuperblockFlags(object):
+    """Probing flags for the superblocks prober."""
+    LABEL    = blkid_c.SUBLKS_LABEL
+    LABELRAW = blkid_c.SUBLKS_LABELRAW
+    UUID     = blkid_c.SUBLKS_UUID
+    UUIDRAW  = blkid_c.SUBLKS_UUIDRAW
+    TYPE     = blkid_c.SUBLKS_TYPE
+    SECTYPE  = blkid_c.SUBLKS_SECTYPE
+    USAGE    = blkid_c.SUBLKS_USAGE
+    VERSION  = blkid_c.SUBLKS_VERSION
+    MAGIC    = blkid_c.SUBLKS_MAGIC
+    BADCSUM  = blkid_c.SUBLKS_BADCSUM
+    DEFAULT  = blkid_c.SUBLKS_DEFAULT
+
+
+class PartitionFlags(object):
+    """Probing flags for the partitions prober."""
+    FORCE_GPT     = blkid_c.PARTS_FORCE_GPT
+    ENTRY_DETAILS = blkid_c.PARTS_ENTRY_DETAILS
+    MAGIC         = blkid_c.PARTS_MAGIC
+
+
+DeviceInfo = collections.namedtuple(
+    'DeviceInfo', ['fstype', 'uuid', 'has_partition']
+)
+
+
+def _need_valid_probe(func):
+    """Ensure that we have a valid probe before calling the function."""
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if self._probe is None:
+            raise ValueError(
+                'trying to call {} with an invalid Probe'.format(func.__name__)
+            )
+        return func(self, *args, **kwargs)
+    return wrapper
+
+
+def _flags_to_bitfield(flags, default=0):
+    if not flags:
+        return default
+    return functools.reduce(lambda x, y: x | y, flags, 0)
+
+
+class Probe(object):
+    @classmethod
+    def from_filename(cls, filepath):
+        """Create a new probe for the device at `filepath`."""
+        probe = blkid_c.new_probe_from_filename(filepath)
+        return cls(probe)
+
+    def __init__(self, probe=None):
+        """Initialize a new probe."""
+        self._probe = blkid_c.new_probe() if probe is None else probe
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
```

Comment:
I'm going to be a bit pedantic, but because this interacts with system-level C stuff, better be safe than sorry.

So, this `probe` thing clearly has a `free_probe` function for cleanup (it's C, after all). However, the current Python interface leaves ways for this release function never to be called:

- I can call `Probe.from_filename` and never use the result in a context manager (i.e., `__exit__` will never be called).
- I can construct a new `Probe` as-is, and never use the result in a context manager.

Since `free_probe` *must* be (promptly) called after `new_probe` is executed, we should change the API to enforce this.

This is, actually, quite simple: instead of making `Probe` a public type, make it 'private' (rename to `_Probe`, only have the `configure` and `probe` members)), then create two top-level functions that enforce use as a context-manager (can use this utility function in `contextlib` with `yield` to implement them, I forgot the name), one `def probe_frome_filename(filepath)` and one `def probe(probe=None)`, and done.

Reply:
Very good point!

I'll rework the interface to enforce correct use.